### PR TITLE
Minimize string copies when rendering.

### PR DESCRIPTION
**Pull request by mbrandonw**
Feels like we could reduce the number of string copies we do under the hood.

### CODE REVIEW

File: Sources/Html/Render.swift [hunk 1]
```diff
@@ -80,3 +57,34 @@ public let voidElements: Set<String> = [
...
+}
+
+private func render(_ attribs: [(String, String?)]) -> String {
+  return attribs
+    .compactMap { key, value in
+      value.map {
+        " " + key + ($0.isEmpty ? "" : "=\"\(escapeAttributeValue($0))\"")
+      }
+    }
+    .joined()
```
Comment (stephencelis) [inline Sources/Html/Render.swift]: Could `append` here too?

File: Sources/Html/Render.swift [hunk 2]
```diff
@@ -80,3 +57,34 @@ public let voidElements: Set<String> = [
...
+
+private func render(_ node: Node, output: inout String) {
+  switch node {
+  case let .comment(string):
+    output.append("<!-- \(escapeHtmlComment(string)) -->")
+  case let .doctype(string):
+    output.append("<!DOCTYPE \(escapeDoctype(string))>")
+  case let .element(tag, attribs, children):
+    let renderedAttribs = render(attribs)
+    let openTag = "<" + tag + renderedAttribs + ">"
```
Comment (stephencelis) [inline Sources/Html/Render.swift]: Worth appending instead of concatenating for the tags?

File: Sources/Html/Render.swift (hunk 1 above)
Comment (mbrandonw) [reply Sources/Html/Render.swift]: yup, done

File: Sources/Html/Render.swift (hunk 2 above)
Comment (mbrandonw) [reply Sources/Html/Render.swift]: yup, done

**Comment by mbrandonw** [comment]
@stephencelis can you give this another look? I ended up just disabling macos tests for now since it isn't playing nicely with xcodegen right now.

**Review by stephencelis — APPROVED** [review]
Looks good!

**Pull request merged**